src/System.Private.CoreLib/shared/System/Runtime/MemoryFailPoint.Windows.cs

```diff
@@ -0,0 +1,110 @@
+        // our last known free address space, hopefully avoiding needing to 
+        // probe again.
+        private static unsafe bool CheckForFreeAddressSpace(ulong size, bool shouldThrow)
+        {
+            // Start walking the address space at 0.  VirtualAlloc may wrap
+            // around the address space.  We don't need to find the exact
+            // pages that VirtualAlloc would return - we just need to
+            // know whether VirtualAlloc could succeed.
+            ulong freeSpaceAfterGCHeap = MemFreeAfterAddress(null, size);
+
+            // Console.WriteLine($"MemoryFailPoint: Checked for free VA space.  Found enough? {(freeSpaceAfterGCHeap >= size)}  Asked for: {size}  Found: {freeSpaceAfterGCHeap}");
+
+            // We may set these without taking a lock - I don't believe
+            // this will hurt, as long as we never increment this number in 
+            // the Dispose method.  If we do an extra bit of checking every
+            // once in a while, but we avoid taking a lock, we may win.
+            LastKnownFreeAddressSpace = (long)freeSpaceAfterGCHeap;
+            LastTimeCheckingAddressSpace = Environment.TickCount;
+
+            if (freeSpaceAfterGCHeap < size && shouldThrow)
+                throw new InsufficientMemoryException(SR.InsufficientMemory_MemFailPoint_VAFrag);
+            return freeSpaceAfterGCHeap >= size;
+        }
+
+        // Returns the amount of consecutive free memory available in a block
+        // of pages.  If we didn't have enough address space, we still return 
+        // a positive value < size, to help potentially avoid the overhead of 
+        // this check if we use a MemoryFailPoint with a smaller size next.
+        private static unsafe ulong MemFreeAfterAddress(void* address, ulong size)
+        {
+            if (size >= s_topOfMemory)
+                return 0;
+
+            ulong largestFreeRegion = 0;
+            Interop.Kernel32.MEMORY_BASIC_INFORMATION memInfo = new Interop.Kernel32.MEMORY_BASIC_INFORMATION();
+            UIntPtr sizeOfMemInfo = (UIntPtr)Marshal.SizeOf(memInfo);
+
+            while (((ulong)address) + size < s_topOfMemory)
+            {
+                UIntPtr r = Interop.Kernel32.VirtualQuery(address, ref memInfo, sizeOfMemInfo);
+                if (r == UIntPtr.Zero)
+                    throw Win32Marshal.GetExceptionForLastWin32Error();
+
+                ulong regionSize = memInfo.RegionSize.ToUInt64();
+                if (memInfo.State == Interop.Kernel32.MEM_FREE)
+                {
+                    if (regionSize >= size)
+                        return regionSize;
+                    else
+                        largestFreeRegion = Math.Max(largestFreeRegion, regionSize);
+                }
+                address = (void*)((ulong)address + regionSize);
+            }
+            return largestFreeRegion;
+        }
+
+        // Allocate a specified number of bytes and commit them. This should enlarge
```

Comment:
> Allocate a specified number of bytes and commit them

... and immediately free them. This method can use a better name, like `GrowPageFileIfNecessaryAndPossible`.

Reply:
I'll go with your suggested name.